lectures/numpy.md

```diff
@@ -475,6 +478,456 @@ A @ (0, 1)
+ax.text(11, 7.0, '?', size=16, ha='center', va='center');
+```
+
+We can see that NumPy cannot expand the arrays to the same size.
+
+It is because, when `b` is expanded from `b -> (3,)` to `b -> (3, 3)`, NumPy cannot match `b` with `a -> (3, 2)`.
+
+Things get even trickier when we move to higher dimensions.
+
+To help us, we can use the following list of rules:
+
+* *Step 1:* When the dimensions of two arrays do not match, NumPy will expand the one with fewer dimensions by adding dimension(s) on the left of the existing dimensions.
+    - For example, suppose `a -> (3, 3)` and `b -> (3,)`, broadcasting will add a dimension to the left so that `b -> (1, 3)`;
+    - Suppose `a -> (2, 2, 2)` and `b -> (2, 2)`, broadcasting will add a dimension to the left so that `b -> (1, 2, 2)`;
+    - Suppose `a -> (3, 2, 2)` and `b -> (2,)`, broadcasting will add two dimensions to the left so that `b -> (1, 1, 2)` (you can also see this process as going through *Step 1* twice).
+
+
+* *Step 2:* When the two arrays have the same dimension but different shapes, NumPy will try to expand dimensions where the shape index is 1.
+    - For example, suppose `a -> (1, 3)` and `b -> (3, 1)`, broadcasting will expand dimensions with shape 1 in both `a` and `b` so that `a -> (3, 3)` and `b -> (3, 3)`;
+    - When `a -> (2, 2, 2)` and  `b -> (1, 2, 2)`, broadcasting will expand the first dimension of `b` so that `b -> (2, 2, 2)`;
+    - When `a -> (3, 2, 2)` and `b -> (1, 1, 2)`, broadcasting will expand `b` on all dimensions with shape 1 so that `b -> (3, 2, 2)`.
+
+Here are code examples for broadcasting higher dimensional arrays
+
+```{code-cell} python3
+# a -> (2, 2, 2) and  b -> (1, 2, 2)
+
+a = np.array(
+    [[[1, 2], 
+      [2, 3]], 
+
+     [[2, 3], 
+      [3, 4]]])
+print(f'the shape of array a is {a.shape}')
+
+b = np.array(
+    [[1,7],
+     [7,1]])
+print(f'the shape of array b is {b.shape}')
+
+a + b
+```
+
+```{code-cell} python3
+# a -> (3, 2, 2) and b -> (2,)
+
+a = np.array(
+    [[[1, 2], 
+      [3, 4]],
+
+     [[4, 5], 
+      [6, 7]],
+
+     [[7, 8], 
+      [9, 10]]])
+print(f'the shape of array a is {a.shape}')
+
+b = np.array([3, 6])
+print(f'the shape of array b is {b.shape}')
+
+a + b
+```
+
+* *Step 3:* After Step 1 and 2, if the two arrays still do not match, a `ValueError` will be raised. For example, suppose `a -> (2, 2, 3)` and `b -> (2, 2)`
+    - By *Step 1*, `b` will be expanded to `b -> (1, 2, 2)`;
+    - By *Step 2*, `b` will be expanded to `b -> (2, 2, 2)`;
+    - We can see that they do not match each other after the first two steps. Thus, a `ValueError` will be raised
+
+```{code-cell} python3
+---
+tags: [raises-exception]
+---
+a = np.array(
+    [[[1, 2, 3], 
+      [2, 3, 4]], 
+     
+     [[2, 3, 4], 
+      [3, 4, 5]]])
+print(f'the shape of array a is {a.shape}')
+
+b = np.array(
+    [[1,7], 
+     [7,1]])
+print(f'the shape of array b is {b.shape}')
+
+a + b
+```
+
 ### Mutability and Copying Arrays
```

Comment:
Currently this section "Mutability and Copying Arrays" appears as a subsection of "Broadcasting", which is not correct.

I suggest that we replace `###` here with `##`.